x-pack/plugins/security_solution/server/lib/telemetry/endpoint_task.ts

```diff
@@ -0,0 +1,197 @@
+  TIMEOUT: '5m',
+  TYPE: 'security:endpoint-meta-telemetry',
+  INTERVAL: '24h',
+  VERSION: '1.0.0',
+};
+
+export class TelemetryEndpointTask {
+  private readonly logger: Logger;
+  private readonly sender: TelemetryEventsSender;
+
+  constructor(
+    logger: Logger,
+    taskManager: TaskManagerSetupContract,
+    sender: TelemetryEventsSender
+  ) {
+    this.logger = logger;
+    this.sender = sender;
+
+    taskManager.registerTaskDefinitions({
+      [TelemetryEndpointTaskConstants.TYPE]: {
+        title: 'Security Solution Telemetry Endpoint Metrics and Info task',
+        timeout: TelemetryEndpointTaskConstants.TIMEOUT,
+        createTaskRunner: ({ taskInstance }: { taskInstance: ConcreteTaskInstance }) => {
+          const { state } = taskInstance;
+
+          return {
+            run: async () => {
+              const lastExecutionTimestamp = moment().utc().toISOString();
+
+              const hits = await this.runTask(taskInstance.id);
+              this.logger.debug(`hits: ${hits}`);
+
+              return {
+                state: {
+                  lastExecutionTimestamp,
+                  runs: (state.runs || 0) + 1,
+                },
+              };
+            },
+            cancel: async () => {},
+          };
+        },
+      },
+    });
+  }
+
+  public start = async (taskManager: TaskManagerStartContract) => {
+    try {
+      await taskManager.ensureScheduled({
+        id: this.getTaskId(),
+        taskType: TelemetryEndpointTaskConstants.TYPE,
+        scope: ['securitySolution'],
+        schedule: {
+          interval: TelemetryEndpointTaskConstants.INTERVAL,
+        },
+        state: { runs: 0 },
+        params: { version: TelemetryEndpointTaskConstants.VERSION },
+      });
+    } catch (e) {
+      this.logger.error(`Error scheduling task, received ${e.message}`);
+    }
+  };
+
+  private getTaskId = (): string => {
+    return `${TelemetryEndpointTaskConstants.TYPE}:${TelemetryEndpointTaskConstants.VERSION}`;
+  };
+
+  public runTask = async (taskId: string) => {
+    if (taskId !== this.getTaskId()) {
+      this.logger.debug(`Outdated task running: ${taskId}`);
+      return 0;
+    }
+
+    const isOptedIn = await this.sender.isTelemetryOptedIn();
+    if (!isOptedIn) {
+      this.logger.debug(`Telemetry is not opted-in.`);
+      return 0;
+    }
+
+    const {
+      body: endpointMetricsResponse,
+    } = ((await this.sender.fetchEndpointMetrics()) as unknown) as {
+      body: EndpointMetricsAggregation;
+    };
+    const endpointMetrics = endpointMetricsResponse.aggregations.endpoint_agents.buckets.map(
+      (epMetrics) => {
+        return {
+          endpoint_agent: epMetrics.latest_metrics.hits.hits[0]._source.agent.id,
+          endpoint_metrics: epMetrics.latest_metrics.hits.hits[0]._source,
+        };
+      }
+    );
+
+    if (endpointMetrics.length === 0) {
+      this.logger.debug('no reported endpoint metrics');
+      return 0;
```

Comment:
Would there ever be a case wherein we have a policy failure but no installed endpoints? Would we want to submit a `endpoint-metadata` payload for that day so that we'd know `Cluster X has an agent with a failed endpoint install`?

Reply:
It is possible - yeah. However, to capture these both in one task run is algorithmically awkward and the security team are going to have a bad time maintaining / scaling it. The reason for this is that we will have a lot of very nested `if`s and conditional branching which is hard to track. If we want to make sure we always capture these we should split these into 2 separate telemetry docs and send to 2 separate channels:

**Metrics / Policy Document**

```js
[
  {
    "agent_id": "****-****-****-****",
    "endpoint_id": "****-****-****-****",
    "metrics": {
      ...
     },
    "policy_config": {
     ...
    } 
  }
]
```

**Policy Repsonse**

```js
[
  {
    "seen_times": 44,
     "config": {
      ...
     },
    "response": {
     ...
    } 
  }
]
```

This would remove a lot of complexity from the code and be easier to maintain. It will also prevent us from sending the same config / response many times - particularly if a users cluster is having a bad day.

Reply:
After meditating on this for a while, I think we can do this in a more efficient way - but it's not going to be much better to maintain. I will implement it this afternoon.